packages/web/src/features/translation/VersesPreview.tsx

```diff
@@ -0,0 +1,150 @@
+import { useQuery } from '@tanstack/react-query';
+import { useEffect, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import apiClient from '../../shared/apiClient';
+import bibleTranslationClient from '../../shared/bibleTranslationClient';
+import { Icon } from '../../shared/components/Icon';
+import { generateReference, isOldTestament, parseVerseId } from './verse-utils';
+import { expandFontFamily } from '../../shared/hooks/useFontLoader';
+import { TextDirection } from '@translation/api-types';
+import LoadingSpinner from '../../shared/components/LoadingSpinner';
+
+type VersesPreviewProps = {
+  language: string;
+  verseIds: string[];
+  onClose: () => void;
+};
+
+function usePreviewQueries(
+  language: string,
+  getContent: boolean,
+  verseIds: string[]
+) {
+  const languagesQuery = useQuery(['languages'], () =>
+    apiClient.languages.findAll()
+  );
+  const translationLanguages = languagesQuery.data?.data ?? [];
+  const selectedLanguage = translationLanguages.find(
+    (l) => l.code === language
+  );
+
+  const originalLanguageQuery = useQuery(
+    ['verse', verseIds.join(','), getContent],
+    async () =>
+      getContent
+        ? Object.fromEntries(
+            await Promise.all(
+              verseIds.map(async (verseId) => [
+                verseId,
+                (await apiClient.verses.findById(verseId)).data.words
+                  .map((w) => w.text)
+                  .join(' '),
+              ])
+            )
+          )
+        : null
+  );
+
+  const translationQuery = useQuery(
+    ['verse-translation', language, verseIds.join(','), getContent],
+    async () =>
+      getContent
+        ? Object.fromEntries(
+            await Promise.all(
+              verseIds.map(async (verseId) => [
+                verseId,
+                await bibleTranslationClient.getTranslation(
+                  verseId,
+                  selectedLanguage?.bibleTranslationIds ?? []
+                ),
+              ])
+            )
+          )
+        : null
+  );
+
+  return {
+    selectedLanguage,
+    originalLanguageQuery,
+    translationQuery,
+  };
+}
+
+export const VersesPreview = ({
+  language,
+  verseIds,
+  onClose,
+}: VersesPreviewProps) => {
+  const { t } = useTranslation(['common', 'bible']);
+  const [title, setTitle] = useState('');
+  const [isValid, setIsValid] = useState(false);
+  const [isHebrew, setIsHebrew] = useState(false);
+
+  useEffect(() => {
+    try {
+      setIsHebrew(isOldTestament(verseIds[0]));
+      setTitle(
+        generateReference(parseVerseId(verseIds[0]), t) +
+          (verseIds.length > 1
+            ? ' - ' +
+              generateReference(parseVerseId(verseIds[verseIds.length - 1]), t)
+            : '')
+      );
+      setIsValid(true);
```

Comment:
All of this is computed state, so ideally, we wouldn't sync this to separate `useState`s, but just compute it inline in the function. If a computed value is expensive to compute, that's what useMemo is for

Reply:
done